internal/jobqueue/fsjobqueue/fsjobqueue.go

```diff
@@ -68,25 +81,38 @@ const channelSize = 100
 // loaded and rescheduled to run if necessary.
 func New(dir string) (*fsJobQueue, error) {
 	q := &fsJobQueue{
-		db:         jsondb.New(dir, 0600),
-		pending:    make(map[string]chan uuid.UUID),
-		dependants: make(map[uuid.UUID][]uuid.UUID),
+		db:           jsondb.New(dir, 0600),
+		pending:      make(map[string]chan uuid.UUID),
+		dependants:   make(map[uuid.UUID][]uuid.UUID),
+		jobIdByToken: make(map[uuid.UUID]uuid.UUID),
+		heartbeats:   make(map[uuid.UUID]time.Time),
 	}
 
 	// Look for jobs that are still pending and build the dependant map.
 	ids, err := q.db.List()
 	if err != nil {
 		return nil, fmt.Errorf("error listing jobs: %v", err)
 	}
+
 	for _, id := range ids {
-		uuid, err := uuid.Parse(id)
+		jobId, err := uuid.Parse(id)
 		if err != nil {
 			return nil, fmt.Errorf("invalid job '%s' in db: %v", id, err)
 		}
-		j, err := q.readJob(uuid)
+		j, err := q.readJob(jobId)
 		if err != nil {
 			return nil, err
 		}
+
+		// If a job is running, and not cancelled, track the token
+		if !j.StartedAt.IsZero() && j.FinishedAt.IsZero() && !j.Canceled {
+			// Ignore older jobs which don't have a token stored
+			if j.Token != uuid.Nil {
```

Comment:
In case the token is not set the job cannot be updated and we should probably fail it on load, shouldn't we? Could be a follow up. Would let you drop the "doesn't work for old jobs" from the news entry ;)

Reply:
+1 for finishing the job.